Split out CPPDEFINES handling in Append methods

mwichmann opened the pull request:
Rather than having lots of special-case code for `CPPDEFINES` in four separate routines, add a new `_add_cppdefines` function to handle it, paying attention to append/prepend, unique/duplicating, and keep-original/replace-original. The existing special case handing was then removed from `Append` and `AppendUnique` (it was never present in `Prepend` and `PrependUnique` anyway - see #3876, but these now get it due to a call to the new function).

Tuple handling is now consistent with list handling: a single tuple is treated as macro names to add, not as a name=value pair. A tuple or list has to be a member of a containing tuple or list to get the macro=value treatment. This *may* affect some existing usage.  macro=value tuples without a value can now be entered either in `(macro,)` form or `(macro, None)` form.

Internally, whenever append/prepend is done, existing contents are forced to a deque, which allows efficient adding at either end without resorting to the tricks the Prepend functions currently do (they still do these tricks, but only in non-`CPPDEFINES` cases).  As a result, values from a dict are not stored as a dict, which has some effect on ordering: values will be *consistently* ordered, but the ones from the dict are no longer necessarily sorted.

In `SCons/Defaults.py`, `processDefines` no longer sorts a dict it is passed, since Python now preserves dict order.  This does not affect the E2E test for `CPPDEFINES` - since those all call an `Append` routine, `CPPDEFINES` will always be a deque, and so processDefines never sees a dict in that case. It could well affect real-life usage - if the setup of CPPDEFINES was such that it used to contain a dict with multiple entries, the order might change (sorting would have presented the keys from that dict in alphabetical order).  This would lead to a one-time rebuild for actions that change (after that it will remain consistent).

In the E2E test CPPDEFINES/append.py some bits were reformatted, and the handler routine now accounts for the type being a deque - since the test does a text comparison of internally produced output, it failed if the word "deque" appeared.  Some new test cases were also added to exercise strings with spaces embedded in them.

Changes were made to the expected output of the E2E test. These reflect changes in the way data is now stored in `CPPDEFINES`, and in some cases in order. Most of these do not change the meaning (i.e. "result" changes, but "final" output is the same). These are the exceptions:

- "appending a dict to a list-of-2lists", `AppendUnique` case: order now preserved as entered (previously took the order of the appended dict)

- "appending a string to a dict", `Append` case: not stored as a dict, so ordering is as originally entered.

- "appending a dict to a dict", `Append` case: no longer merge into a dict, so this is now an actual append rather than a merge of dicts which caused the uniquing effect even without calling `AppendUnique` (arguably the old way was incorrect).

A new `test/CPPDEFINES/prepend.py` is added to test `Prepend*` cases.

`append.py` and `prepend.py` are structured to fetch the SConstruct from a fixture file.

`append.py` got an added test in the main text matrix, a string of the macro=value form. The same 5x5 maxtrix is used in the new `prepend.py` test as well ("expected" values for these had to be added as well).

Cosmetically, append and prepend now print their test summary so strings have quotation marks - the "orig" lines in the expected output was adjusted.  This change looks like:

```txt
-   orig = FOO, append = FOO
+   orig = 'FOO', append = 'FOO'
```

The other tests in `test/CPPDEFINES` got copyright updating and reformatting, but otherwise do not change.

Documentation updated to clarify behavior.

Fixes #4254
Fixes #3876

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `CHANGES.txt` (and read the `README.rst`)
* [X] I have updated the appropriate documentation

bdbaddog commented on RELEASE.txt:
```diff
@@ -64,6 +69,17 @@ FIXES
 - Ninja: Fix execution environment sanitation for launching ninja. Previously if you set an
   execution environment variable set to a python list it would crash. Now it
   will create a string joining the list with os.pathsep
+- Inconsistent behavior of adding CPPDEFINES values via tuple arguments
```
Since this can change the CPPDEFINES, we should add a section at the top with
NOTE: The following usages of CPPDEFINES have changed
env.Append(CPPDEFINES,('a','b') )  now yields -Da -Db instead of -Da=b
etc..
Should be a fairly small set of these to clearly spell out what's changed?

bdbaddog commented on CHANGES.txt:
```diff
@@ -86,7 +86,17 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
       tests are now run on changes in this area so future problems can
       be spotted.
-
+    - Special-case handling for CPPDEFINES in the Append/Prepend routines
```
This should probably be multiple bullets.

mwichmann replied:
Can do that, but it's really convouted - see the long report in the one issue of the various inconsistencies I had discovered.  So there would be a whole bunch of "now get (b) instead of (a), but only if (c); if (d) the behavior was already (b); if (e), this previously did not work at all".

bdbaddog replied:
I think that would be valuable.  This whole PR is really about fixing some corner cases (and reorganizing the code so it's more sane), so explaining which corner cases are now fixed (changed) would be useful for anyone who updated and saw their build change, and/or has a build not doing what they expect and sees that the newer version solves that.

Bigger organizations often need a forcing function to move to a newer version.

bdbaddog commented on SCons/Defaults.xml:
```diff
@@ -92,68 +92,111 @@ to each definition in &cv-link-CPPDEFINES;.
+If any element is a tuple (or list)
+then the first item of the tuple is the macro name
+and the second is the macro definition.
+If the definition is not omitted or <literal>None</literal>,
```
value instead of definition?

mwichmann replied:
I thought I was writing it in terms of how the C standard expresses those... although now I look, the standard uses "replacement" rather than "definition".

bdbaddog commented on SCons/Defaults.xml:
```diff
@@ -92,68 +92,111 @@ to each definition in &cv-link-CPPDEFINES;.
+</example_commands>
+
+<para>
+:<emphasis>Changed in version  4.5</emphasis>:
```
4.5.0

bdbaddog commented on SCons/Environment.xml:
```diff
@@ -499,134 +499,199 @@ Multiple targets can be passed in to a single call to
-CXXFLAGS=-std=c11 -O, CPPDEFINES=['RELEASE', 'EXTRA']
-CPPDEFINES will expand to -DRELEASE -DEXTRA
+CXXFLAGS = -std=c11, CPPDEFINES = RELEASE
+CXXFLAGS = -std=c11 -O, CPPDEFINES = deque(['RELEASE', 'EXTRA'])
```
adding deque here will most likely just confuse users. Many of whom aren't strong python programmers. I'd leave it just the list.

bdbaddog commented on SCons/Environment.xml:
```diff
@@ -499,134 +499,199 @@ Multiple targets can be passed in to a single call to
+(<computeroutput>-DBAR=1</computeroutput>).
+&SCons; allows you to specify a macro with a definition
+using a <literal>name=value</literal> string,
+or a tuple <literal>(name, value)</literal>
```
Can this no longer be done with a list?

mwichmann replied:
It can, it's just awkward to keep repeating multiples all over the place. It would be nice to use "sequence" as that covers all of tuple, list, UserList, etc. but it also includes str, which is not a data type that can be used for a name, value pair.

mwichmann replied:
I pushed a rewording of this part.

bdbaddog commented on SCons/Environment.xml:
```diff
@@ -499,134 +499,199 @@ Multiple targets can be passed in to a single call to
-CPPDEFINES=['FOO', 'BAR=1', ('OTHER', 2)]
-CPPDEFINES=['FOO', 'BAR=1', ('OTHER', 2), {'EXTRA': 'arg'}]
+CPPDEFINES = FOO
+CPPDEFINES = deque(['FOO', 'BAR=1'])
```
deque?

mwichmann replied:
The code takes the initial value given to CPPDEFINES as-is, but once you further Append/Prepend it takes the liberty of converting to a type useful to it.  It always did that, but the chosen compound type depends on the sequence of events - could be a list, a dict, etc. Or in the case of AppendUnique, it converted it to a type incorrectly in some cases. Now we consistently get a deque, allowing the processing to be sane. Yes, as noted elsewhere, it could be a list instead but that shouldn't really be important to the user.  Which type is really only exposed if you dig, as it happens here trying to show an example.

bdbaddog commented on SCons/Environment.xml:
```diff
@@ -499,134 +499,199 @@ Multiple targets can be passed in to a single call to
+</screen>
+
+<para>
+<emphasis>Changed in version  4.5</emphasis>:
```
4.5.0

bdbaddog commented on SCons/Environment.xml:
```diff
@@ -499,134 +499,199 @@ Multiple targets can be passed in to a single call to
+<screen>
+$ scons -Q
+CPPDEFINES = ('MACRO1', 'MACRO2')
+CPPDEFINES = deque(['MACRO1', 'MACRO2', ('MACRO3', 'MACRO4')])
```
deque?

bdbaddog commented:
Curious: Why use deque instead of a list to hold the defines?

bdbaddog commented on SCons/Environment.py:
```diff
@@ -193,6 +193,162 @@ def _delete_duplicates(l, keep_last):
+        """Returns match for *item* if found in *defines*.
+
+        Accounts for type differences: ("FOO", "BAR"), ["FOO", "BAR"]
+        "FOO=BAR" and {"FOO": "BAR"} all iffer as far as Python equality
```
all differ (instead of iffer)

mwichmann replied:
typo, yeah.

bdbaddog commented on SCons/Environment.py:
```diff
@@ -193,6 +193,162 @@ def _delete_duplicates(l, keep_last):
+        delete_existing: if *unique* is true, add *val* after removing previous.
+    """
+
+    def _add_define(item, defines: deque, prepend: bool = False) -> None:
```
not a fan of embedded functions..

mwichmann replied:
What can I say?  They have their uses. Obviously, this one can be unrolled. I think it makes it a bit more readable.

mwichmann commented:
Prepend is as fast as append without having to play silly games (reverse, append reverse... that kind of thing)

mwichmann commented:
> Curious: Why use deque instead of a list to hold the defines?

This is a natural fit for a deque since it's designed to be added to from either end with Prepend and Append. Prepending a deque is as fast as appending, and this way don't have to play silly games (reverse the list, append to it, reverse back... that kind of thing).

bdbaddog commented:
Would you expect all the examples in this sample code https://github.com/bdbaddog/scons-bugswork/tree/main/4254
(Derived from description in Issue #4254 )
To work with your current code here? I'm seeing some issues.

This is the example run on current master:
```
% python ~/devel/scons/git/as_scons/scripts/scons.py -n
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
gcc -o main_case0.o -c -Dz=case2 main.c
gcc -o main_case1.o -c "-D('z', 'case1')" main.c
gcc -o main_case2.o -c -Dz=case2 -Da -Db main.c
gcc -o main_case4.o -c -Dcase4 -Dz=case4 main.c
gcc -o main_case5.o -c -Dz -Dcase5 main.c
scons: done building targets.
```

This is the example run against this branch (Note FIXED=1 enables case 3 which crashes in master)

```
% python ~/devel/scons/git/mwichmann/scripts/scons.py -n FIXED=1
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
gcc -o main_case0.o -c -Dz=case2 main.c
gcc -o main_case1.o -c "-D('z', 'case1')" main.c
gcc -o main_case2.o -c -Dz -Dcase2 -Da -Db main.c
gcc -o main_case3.o -c -Dz -Dcase3 -Dsimple_string main.c
gcc -o main_case4.o -c -Dcase4 main.c
gcc -o main_case5.o -c -Dz -Dcase5 main.c
scons: done building targets.
```

mwichmann commented:
I've pushed a rebase of this (had to force), dealing with Util having been refactored meanwhile. A couple of small changes, plus test expansion to make sure those cases are covered, coming next. Then back to the RELEASE/CHANGES material.

mwichmann commented:
FWIW, this is the output from the current version:
```
scons: Building targets ...
gcc -o main_case0.o -c -Dz=case2 main.c
gcc -o main_case1.o -c -Dz -Dcase1 main.c
gcc -o main_case2.o -c -Dz -Dcase2 -Da -Db main.c
gcc -o main_case3.o -c -Dz -Dcase3 -Dsimple_string main.c
gcc -o main_case4.o -c -Dcase4 -Dz -Dcase4 main.c
gcc -o main_case5.o -c -Dz -Dcase5 main.c
scons: done building targets.
```

mwichmann commented:
Define *correct* here. If you specifically wanted the output result `-Dz=case2` the three ways to enter it are:
```py
[('z','case2')]
{'z': 'case2'}
"z=case2"
```
Only your case0 uses one of those three forms.

Otherwise, I'm not sure what you're after... the appending is all being handled as I expect now.

mwichmann commented:
Just to reiterate a point from elsewhere, this has *always* been ambiguous:
```py
case_4.Append(CPPDEFINES=('z', 'case4'))
```
is that the same as
```py
case_4.Append(CPPDEFINES=['z', 'case4'])
```
or is tuple interpreted differently than list in this case? the new behavior is the two work the same - you asked to add two separate macros to CPPDEFINES.  previously it worked all kinds of different ways - depending on which method's special-case-for-CPPDEFINES code you were in.

bdbaddog commented:
I sit corrected.. ;)
Indeed CPPDEFINES=('a','b') spitting out -Da -Db matches the docs.

mwichmann commented:
Just to overload this thing further, here's the output of a simple demo: starting with a `CPPDEFINES` value of the string `"FOO"`, in a loop we add each of five different ways that *could* be a valued-macro (tuple, list, tuple-in-a-list, string, dict) using each of the four functions.
```py
(venv-system) $ git switch -
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
(venv-system) [mats@boulder scons]$ scons -Q -f exp/more-4263/SConstruct
'FOO' Append ('B', 'DEBUG') -> ['FOO', 'B=DEBUG']
'FOO' Append ['B', 'DEBUG'] -> ['FOO', 'B', 'DEBUG']
'FOO' Append [('B', 'DEBUG')] -> ['FOO', 'B=DEBUG']
'FOO' Append 'B=DEBUG' -> ['FOO', 'B=DEBUG']
'FOO' Append {'B': 'DEBUG'} -> ['FOO', 'B=DEBUG']
'FOO' AppendUnique ('B', 'DEBUG') EXCEPTION: can only concatenate list (not "tuple") to list
'FOO' AppendUnique ['B', 'DEBUG'] -> ['FOO', 'B', 'DEBUG']
'FOO' AppendUnique [('B', 'DEBUG')] -> ['FOO', 'B=DEBUG']
'FOO' AppendUnique 'B=DEBUG' -> ['FOO', 'B=DEBUG']
'FOO' AppendUnique {'B': 'DEBUG'} -> ['FOO', 'B=DEBUG']
'FOO' Prepend ('B', 'DEBUG') EXCEPTION: 'str' object has no attribute 'insert'
'FOO' Prepend ['B', 'DEBUG'] -> ['B', 'DEBUG', 'FOO']
'FOO' Prepend [('B', 'DEBUG')] -> ['B=DEBUG', 'FOO']
'FOO' Prepend 'B=DEBUG' -> ['B=DEBUGFOO']
'FOO' Prepend {'B': 'DEBUG'} EXCEPTION: 'str' object has no attribute 'insert'
'FOO' PrependUnique ('B', 'DEBUG') EXCEPTION: can only concatenate tuple (not "str") to tuple
'FOO' PrependUnique ['B', 'DEBUG'] -> ['B', 'DEBUG', 'FOO']
'FOO' PrependUnique [('B', 'DEBUG')] -> ['B=DEBUG', 'FOO']
'FOO' PrependUnique 'B=DEBUG' -> ['B=DEBUGFOO']
'FOO' PrependUnique {'B': 'DEBUG'} EXCEPTION: unsupported operand type(s) for +: 'dict' and 'str'
scons: `.' is up to date.
(venv-system) $ git switch -
Switched to branch 'maint/CPPDEFINES-Append'
Your branch is up to date with 'origin/maint/CPPDEFINES-Append'.
(venv-system) $ scons -Q -f exp/more-4263/SConstruct
'FOO' Append ('B', 'DEBUG') -> ['FOO', 'B', 'DEBUG']
'FOO' Append ['B', 'DEBUG'] -> ['FOO', 'B', 'DEBUG']
'FOO' Append [('B', 'DEBUG')] -> ['FOO', 'B=DEBUG']
'FOO' Append 'B=DEBUG' -> ['FOO', 'B=DEBUG']
'FOO' Append {'B': 'DEBUG'} -> ['FOO', 'B=DEBUG']
'FOO' AppendUnique ('B', 'DEBUG') -> ['FOO', 'B', 'DEBUG']
'FOO' AppendUnique ['B', 'DEBUG'] -> ['FOO', 'B', 'DEBUG']
'FOO' AppendUnique [('B', 'DEBUG')] -> ['FOO', 'B=DEBUG']
'FOO' AppendUnique 'B=DEBUG' -> ['FOO', 'B=DEBUG']
'FOO' AppendUnique {'B': 'DEBUG'} -> ['FOO', 'B=DEBUG']
'FOO' Prepend ('B', 'DEBUG') -> ['DEBUG', 'B', 'FOO']
'FOO' Prepend ['B', 'DEBUG'] -> ['DEBUG', 'B', 'FOO']
'FOO' Prepend [('B', 'DEBUG')] -> ['B=DEBUG', 'FOO']
'FOO' Prepend 'B=DEBUG' -> ['B=DEBUG', 'FOO']
'FOO' Prepend {'B': 'DEBUG'} -> ['B=DEBUG', 'FOO']
'FOO' PrependUnique ('B', 'DEBUG') -> ['DEBUG', 'B', 'FOO']
'FOO' PrependUnique ['B', 'DEBUG'] -> ['DEBUG', 'B', 'FOO']
'FOO' PrependUnique [('B', 'DEBUG')] -> ['B=DEBUG', 'FOO']
'FOO' PrependUnique 'B=DEBUG' -> ['B=DEBUG', 'FOO']
'FOO' PrependUnique {'B': 'DEBUG'} -> ['B=DEBUG', 'FOO']
scons: `.' is up to date.
(venv-system) $
```
These kind of tests happen in `test/CPPDEFINES/append.py` and in the new `test/CPPDEFINES/prepend.py` - the prepend behavior with CPPDEFINES was never tested - this is just a subset.

Yes, at the very top, appending a tuple had the valued-macro effect, while appending a list didn't; now they work the same. We can restore that behavior, as long as we document it (recall in the CPPDEFINES docs, it did not distinguish between tuple and list in describing how such an element was handled, so why should there be a difference when appending?). append-unique, prepend, and prepend-unique all took an exception if you tried to add a tuple. You can see some other amusing problems (leading one to believe probably nobody ever tried these) - prepending the string form botches the result; prepending the dict form takes an exception.

bdbaddog commented:
Updated logic to match current behavior when you append a tuple to yield `-Dkey=value`

bdbaddog commented:
The only question mark is, do we change specifying CPPDEFINES initially as a tuple to yield `-Dkey=value`, or leave as is.
It would seem more consistent if it was changed.

Thoughts?

mwichmann commented on RELEASE.txt:
```diff
@@ -17,6 +17,9 @@ NOTE: If you build with Python 3.10.0 and then rebuild with 3.10.1 (or higher),
       see unexpected rebuilds. This is due to Python internals changing which changed
       the signature of a Python Action Function.
 
+NOTE: If you use dictionaries to specify your CPPDEFINES, most likely the order of the CPPDEFINES
```
That note, basically, was already in place. Okay if you think it's helpful to have it at the top as well.

bdbaddog replied:
We always put the "it may rebuild" warnings at the top.  to try to minimize  the TL;DR issue...

mwichmann commented:
If you're okay with tuple specifically having a different meaning than list - personally, I think *that* is inconsistent. If so, then some further rework is needed (see below).  What do you propose happens if someone tried to Append, say, `('FOO', ('BAR', '22'))`?  Currently that would end up expanding to `"FOO=('BAR', '22')"`, whereas a list version of the same becomes `['FOO', 'BAR=22']`.

Rework:
* docs need to describe that adding a tuple by itself becomes a valued-macro
* the test cases will need to be gone through to reset the expectations for the changed behavior.
* "initial value" needs to be handled. With the latest change, Appending a tuple when `CPPDEFINES` doesn't exist yet gives a different result than when `CPPDEFINES` has any value.  Obviously, that's a no-go - user code isn't supposed to have to know if it already had a value. This happens in `processDefines`.
* currently, `processDefines` takes *any* sequence embedded in a sequence as a valued-macro. What should it do if it sees a non-tuple here? This is the pre-PR versionn (mine only combines the two conditions into a single `is_Sequence` check):
 ```py
            elif SCons.Util.is_List(d) or isinstance(d, tuple):
                if len(d) >= 2:
                    l.append(str(d[0]) + '=' + str(d[1]))
                else:
                    l.append(str(d[0]))
```

mwichmann commented:
Attaching the result of running the CPPDEFINES tests as they exist in the PR against a version that changes both the add-single-tuple as well as the `processDefines`-single-tuple behavior.

[Changed_tuple.txt](https://github.com/SCons/scons/files/10581306/Changed_tuple.txt)

mwichmann commented:
Even with the two changes there's still a hole: appending a tuple works as you want, and initializing to a single tuple works (because then `processDefines just sees that tuple`), but initializing with a tuple and then appending (anything) to it does not (because at the time of the first append the original tuple is converted into the internal datatype first, losing its tuple status). You want me to keep going down this road?

mwichmann commented:
One more question is what to do with strings - when to split on space and when not.  We know a string entry *inside* a list is promised not to be split, so you can enter entries with spaces (that's not CPPDEFINES specific, of course). But:
```
initial 'FOO BAR=22' without adds -> ['FOO BAR=22']  # initial is not split if never added to
iniitial 'FOO BAR=22' Append ('B', 'DEBUG') -> ['FOO', 'BAR=22', 'B=DEBUG']  # initial has been split
initial None Append 'BAZ B=DEBUG' -> ['BAZ', 'B=DEBUG']  # addition has been split
initial 'FOO BAR=22' Append 'BAZ B=DEBUG' -> ['FOO', 'BAR=22', 'BAZ B=DEBUG']  # initial split, addition has not been split
```

Isn't this fun?

bdbaddog commented:
> Even with the two changes there's still a hole: appending a tuple works as you want, and initializing to a single tuple works (because then `processDefines` just sees that tuple`), but initializing with a tuple and then appending (anything) to it does not (because at the time of the first append the original tuple is converted into the internal datatype first, losing its tuple status). You want me to keep going down this road?

Seems like a tuple should always defined `-Dkey=value` and not change when somethings appended to it right?

bdbaddog commented:
oopsie.. accidently closed.

mwichmann commented:
> Seems like a tuple should always defined `-Dkey=value` and not change when somethings appended to it right?

If the decision is that's what a tuple is, then it should be consistently so - see my quote from earlier:
> Obviously, that's a no-go

mwichmann commented:
oops, spotted that I'd made an error in preparing the Prepend test (copy-paste error, hadn't changed one instance of Append to Prepend), and fixed it... and then didn't remember to go fix the "expected" text.  Fix for that forthcoming.
